### PR TITLE
Features

### PR DESCRIPTION
**Pull request by 2HgO**
- Ability to create custom serializer/deserializer
- Ability to skip field using omitempty pragma
- remove unknown type `ClientKind` from single.nim
- add tests for custom serialization/deserialization
- add tests for omitting fields using omitempty pragma

### CODE REVIEW

**Comment by 2HgO** [comment]
@disruptek please help check this out :)

File: mongo/bson.nim [hunk 1]
```diff
@@ -809,17 +831,19 @@ proc to*(b: Bson, T: typedesc): T =
       var key = k
       when val.hasCustomPragma(dbKey):
         key = val.getCustomPragmaVal(dbKey)
-      if key notin b:
+      if key notin b and not val.hasCustomPragma(omitempty):
         raise BsonError.newException "Key " & key & " not found for " & $T
-      val = b[key].to(type(val))
+      if key in b:
+        val = b[key].to(type(val))
```
Comment (disruptek) [inline mongo/bson.nim]: Please change these to a single if/elif statement; `key in b` and then `elif no omit-empty` form.

File: mongo/bson.nim [hunk 2]
```diff
@@ -767,10 +767,32 @@ proc newBsonDocument*(bytes: string): Bson =
...
+    ## type that can serialize itself to a bson object (implemented by all primitive type)
+    s.toBson() is Bson
+  BsonDeserializer* = concept d
+    ## type that can deserialize a bson type to itself
+    Bson.to(d)
+
 template dbKey*(name: string) {.pragma.}
+template omitempty*() {.pragma.}
+
+proc toBson*(d: DateTime) : Bson = ## Serializer for DateTime type
```
Comment (disruptek) [inline mongo/bson.nim]: style: Documentation comments should go on their own line when they are the first statement of a procedure body.

File: mongo/bson.nim [hunk 3]
```diff
@@ -767,10 +767,32 @@ proc newBsonDocument*(bytes: string): Bson =
 
 ## Serialization/deserialization ext
 
+# ------------- type: Serializer/Deserializer -------------------#
+
+type
+  BsonSerializer* = concept s
+    ## type that can serialize itself to a bson object (implemented by all primitive type)
```
Comment (disruptek) [inline mongo/bson.nim]: _all primitive types_

File: mongo/bson.nim [hunk 4]
```diff
@@ -809,17 +831,19 @@ proc to*(b: Bson, T: typedesc): T =
...
     for k, val in fieldPairs(result):
       var key = k
       when val.hasCustomPragma(dbKey):
         key = val.getCustomPragmaVal(dbKey)
-      if key notin b:
+      if key notin b and not val.hasCustomPragma(omitempty):
         raise BsonError.newException "Key " & key & " not found for " & $T
-      val = b[key].to(type(val))
+      if key in b:
+        val = b[key].to(type(val))
```
Comment (disruptek) [inline mongo/bson.nim]: Please update uses of `type` to modern `typeof`; the former is deprecated for this purpose.

File: mongo/bson.nim [hunk 5]
```diff
@@ -841,9 +865,13 @@ proc toBson*[T](entry: T): Bson =
     result = newBsonDocument()
     for k, v in fieldPairs(entry):
       when v.hasCustomPragma(dbKey):
-        result[v.getCustomPragmaVal(dbKey)] = toBson(v)
+        when v.hasCustomPragma(omitempty):
+          if default(type(v)) != v: result[v.getCustomPragmaVal(dbKey)] = toBson(v)
```
Comment (disruptek) [inline mongo/bson.nim]: style: Please resist single-line `if` statements.

**Review by disruptek — CHANGES_REQUESTED** [review]
Thanks for using this stuff -- did you run into any other problems?

Concepts are in a difficult state; the original implementation is broken and doesn't compose for any bug the most trivial uses.  There's a replacement implementation but in my opinion, the replacement misses most of the ADT abstraction from the original implementation.

If you aren't bothered by this, I'm not bothered by this. :wink:

**Comment by 2HgO** [comment]
> 

I didn't run into any problems working with concepts. I believe this usecase is simple enough that it will not introduce any unwanted behaviours.

**Review by disruptek — APPROVED** [review]
Thank you!

**Pull request merged**